Initial WireGuard driver implementation

deanlorenz opened the pull request.

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	var cidr *net.IPNet
+	for _, sn := range remoteEndpoint.Spec.Subnets {
+		if _, cidr, err = net.ParseCIDR(sn); err != nil {
+			return fmt.Errorf("failed to parse subnet %s: %v", sn, err)
```
Perhaps just log the error and continue? Either way we shouldn't get a bad CIDR...

deanlorenz replied:
Done, ignoring malformed subnets. Might create peers with no allowed IPs.

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+		}
+	}
+
+	klog.V(log.TRACE).Infof("Done removing endpoint for cluster %s", remoteEndpoint.Spec.ClusterID)
```
I notice you use TRACE level throughout. Any particular reason to not use DEBUG? We expect DEBUG to be the norm with TRACE only used for really noisy logs or ones that may dump a lot of data.

deanlorenz replied:
Moved all to DEBUG

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+		Peers:        peerCfg,
+	}); err != nil {
+		return fmt.Errorf("failed to remove wireguard peer with key %s: %v", remoteKey.String(), err)
+	}
```
L322-332 are largely duplicated in `ConnectToEndpoint` - please create a function.

deanlorenz replied:
done

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	if err = w.client.ConfigureDevice(DefaultDeviceName, wgtypes.Config{
+		ReplacePeers: false,
+		Peers:        peerCfg,
+	}); err != nil {
```
Please separate lines to improve readability.

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	}
+
+	klog.V(log.TRACE).Infof("Successfully connected endpoint peer %+v", peerCfg)
+
```
Can you split this into separate functions where appropriate?

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	// Add routes to peer
+	//TODO save old routes for removal
+	var wg netlink.Link
+	if wg, err = netlink.LinkByName(DefaultDeviceName); err != nil {
```
Can't use use `w.link` instead of retrieving it again?

deanlorenz replied:
done

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+
+	// del routes
+	var wg netlink.Link
+	if wg, err = netlink.LinkByName(DefaultDeviceName); err != nil {
```
Can't use use `w.link` instead of retrieving it again?

deanlorenz replied:
done

tpantelis commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+		Peers:        peerCfg,
+	}); err != nil {
+		return fmt.Errorf("failed to remove wireguard peer with key %s: %v", remoteKey.String(), err)
+	}
```
I'm not clear on why you proceed with removing the peer if the keys didn't match. Is this case expected under certain circumstances? Can you add comments to explain this?

deanlorenz replied:
added comment in code to explain meaning of key mismatch

mangelajo commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+package wireguard
```
being a nit, but I agree, we try to avoid repetition, this would be 

cable/wireguard/driver , quite meaningful :)

deanlorenz replied:
done

mangelajo commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+}
+
+// NewWGDriver creates a new Wireguard driver
+func NewWGDriver(localSubnets []string, localEndpoint types.SubmarinerEndpoint) (cable.Driver, error) {
```
:+1: as when used in code it would read wireguard.NewDriver(...)

deanlorenz replied:
done

mangelajo commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	}
+
+	klog.V(log.TRACE).Infof("Initialized wireguard %s with publicKey %s", DefaultDeviceName, pub.String())
+	return &wg, nil
```
:+1:

deanlorenz replied:
done

deanlorenz commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	var remoteKey wgtypes.Key
+	if remoteKey, err = wgtypes.ParseKey(key); err != nil {
+		return fmt.Errorf("failed to parse public key %s: %v", key, err)
+	}
```
done

deanlorenz commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+	if err = w.client.ConfigureDevice(DefaultDeviceName, wgtypes.Config{
+		ReplacePeers: false,
+		Peers:        peerCfg,
+	}); err != nil {
```
done

deanlorenz commented on pkg/cable/wireguard/WGdriver.go:
```diff
@@ -0,0 +1,365 @@
+		klog.Warningf("Key mismatch for peer cluster %s, keeping existing routes", remoteEndpoint.Spec.ClusterID)
+		return nil
+	}
+	delete(w.peers, remoteEndpoint.Spec.ClusterID)
```
Added comments in code. Cannot delete peers entry until checking mismatch -- clusterID may have a valid entry with a different key.

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+}
+
+// Create new wg link and assign addr from local subnets
+func setWGLink(w *wireguard, localSubnets []string) error {
```
minor thing but you could `w *wireguard` the receiver instead of explicitly passing in.

deanlorenz replied:
done

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+	for _, a := range addrs {
+		ip, _, err := net.ParseCIDR(a.String())
+		if err != nil {
+			klog.V(log.DEBUG).Infof("Skipping local address %v, unable to ParseCIDR: %v", a, err)
```
It seems this should be logged as error or warning as we do on L365? I assume this isn't expected...

deanlorenz replied:
I don't have a strong opinion either way, but the errors are slightly different. 
L365 is submariner error -- malformed subnets -- we can continue but something is probably not working as sub CIDRs will not be added to allowed IPs and not routed to other cluster.
L342 is netlink library error or some weird device that returns an unexpected address -- we simply don't use it and look for another internal address. Only if we cannot find any address this becomes an error at L355.

tpantelis replied:
👍 makes sense.

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+
+	// parse localSubnets and get internal address
+	w.localSubnets = parseSubnets(localSubnets)
+	ip, err := discoverInternalIP(w.localSubnets)
```
What if `discoverInternalIP` returned an error indicating it wasn't able to determine if an internal address exists? Should we still proceed with using the endpoint IP?

deanlorenz replied:
that is what is attempted at L311. 
Using endpoint IP will probably cause problems for pods on host network.

tpantelis replied:
L311 is a log statement - did you mean L314? 😄 The call site here can't distinguish between an internal address not found and an error trying to determine if an internal address exists from `discoverInternalIP`.  For the former,  we certainly want to proceed to L309 to use the endpoint IP.  I'm unclear about the latter but if we still want to proceed with attempting to use the endpoint IP in this case as well then I think we should at least log the error. I propose to return nil from `discoverInternalIP` if no address is found and log the returned error here, ie:

```
ip, err := discoverInternalIP(w.localSubnets)
if err != nil {
    klog.Errorf("Error attempting to discover an internal address: %v", err)
}

if ip == nil {
    klog.V(log.DEBUG).Infof("Using endpoint IP as internal address; %v", err)
    ip = endpointIP(&w.localEndpoint)
}

```

deanlorenz replied:
Actually L309 (removed 2 comments above...).
I see your point; however, `discoverInternalIP` returns `""` in both cases.
I can change `discoverInternalIP` to only log the 'not found' case, but I still see both cases as error, and still think the best thing we can do is try the ep IP.

tpantelis replied:
> Actually L309 (removed 2 comments above...).
> I see your point; however, `discoverInternalIP` returns `""` in both cases.

yeah, sorry, return `""` instead of nil.

> I can change `discoverInternalIP` to only log the 'not found' case, but I still see both cases as error, and still think the best thing we can do is try the ep IP.

yeah try the endpoint IP in both cases - I'm just suggesting to log the error if something failed preventing the determination so we don't lose it in case it might be helpful for troubleshooting. So the last line in `discoverInternalIP`  where it's not found  would return  `"", nil`.

deanlorenz replied:
done

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+		_, cidr, err := net.ParseCIDR(sn)
+		if err != nil {
+			// this should not happen. Log and continue
+			klog.Errorf("failed to parse subnet %s: %v", sn, err)
```
Should continue here and not fall thru to L367.

deanlorenz replied:
oops. good catch. Done.

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+	})
+	if err != nil {
+		klog.Errorf("Failed to remove WireGuard peer with key %s: %v", key, err)
+		return err
```
Since the call sites ignore the returned error, I'd prefer to not return it. I think call sites look cleaner w/o  having to suppress it with `_`.

deanlorenz replied:
The error path at the caller is currently very basic we may want to improve it later. I find that ignoring an error explicitly indicates that. I'd rather have the utility return the error as a failure indication. 
I actually find it cleaner as is, but if you prefer I can remove it.

tpantelis replied:
I understand what you're saying. It makes sense to return an error and not assume the caller will ignore it, although we are kind of assuming that by logging the error on L386 😄 It's fine the way it is.

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+	client        *wgctrl.Client
+	link          netlink.Link
+	//debug   bool
+	//logFile string
```
Please remove commented lines.

deanlorenz replied:
done

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,411 @@
+
+	// verify peer was added
+	// TODO verify configuration
+	if p, err := peerByKey(w.client, remoteKey); err != nil {
```
Shouldn't success from `ConfigureDevice` above be enough to know that the peer was added?

deanlorenz replied:
It should, but we had some silent failures in our tests. So added this for better debugging.
Peer configurations is not that common, so I don't think it is a problem to verify.

tpantelis commented:
Can you use another PR for https://github.com/submariner-io/submariner/pull/441/commits/4a015470685b53b66a1c2121493c31410e103b7b? The e2e scripting is being refactored right now with most of it moving to the shipyard project so it's in a holding pattern for new features right now until that work is done.

deanlorenz commented:
Added option to make e2e to allow setting wireguard as cable driver:
`make e2e cable_driver=wireguard`
Removed code to set wireguard as default driver

@mangelajo the make option did not work for me without changing `scripts/e2e` to `kind-e2e/e2e.sh` (not use `/opt/`. Even `make dapper-image` did not help.

deanlorenz commented:
@tpantelis reverted the make change and added PR [#476](https://github.com/submariner-io/submariner/pull/476)

tpantelis commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,416 @@
+			}
+		}
+	}
+	klog.Warningf("could not find internal address in local subnets")
```
```suggestion
	klog.Warningf("Could not find an internal address in %v that matches a local subnet in %v", addrs, cidrs)
```

Sorry one more request - since this is a warning, it seems it would be useful to include some context for possible troubleshooting. If you think otherwise then we can leave as is.

mkolesnik commented on pkg/cable/wireguard/driver.go:
```diff
@@ -0,0 +1,416 @@
+	PublicKey = "publicKey"
+
+	// we assume Linux
+	//deviceType = wgtypes.LinuxKernel
```
This needed?

deanlorenz replied:
This was a placeholder for user-space wireguard support. 
I removed for now and will add comment to README

mkolesnik commented on pkg/cable/wireguard/README.md:
````diff
@@ -0,0 +1,53 @@
+
+- The driver needs to be enabled with
+  ```ShellSession
+  $ bin/subctl join --kubeconfig wg3-conf --cable-driver wireguard --disable-nat  --repository roytman --version  latest broker-info.subm
````
I'd drop the repo & version flags (unless they have some significance specifically for WG?)

deanlorenz replied:
They are no longer significant, as latest subctl now supports the --cable-driver option. Removed.
